prow/plugins/wip/wip-label.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package wip will label a PR a work-in-progress if the author provides
+// a prefix to their pull request title to the same effect. The submit-
+// queue will not merge pull requests with the work-in-progress label.
+// The label will be removed when the title changes to no longer begin
+// with the prefix.
+package wip
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	label      = "do-not-merge/work-in-progress"
+	pluginName = "wip"
+)
+
+var (
+	titleRegex = regexp.MustCompile(`(?i)^(\[WIP\]|WIP)`)
+)
+
+type event struct {
+	org         string
+	repo        string
+	number      int
+	hasLabel    bool
+	needsLabel  bool
+	commentBody string
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	CreateComment(org, repo string, number int, comment string) error
+	DeleteComment(org, repo string, ID int) error
+}
+
+func handlePullRequest(pc plugins.PluginClient, pe github.PullRequestEvent) error {
+	// These are the only actions indicating the PR title may have changed.
+	if pe.Action != github.PullRequestActionOpened && pe.Action != github.PullRequestActionEdited && pe.Action != github.PullRequestActionSynchronize {
```

Comment:
I don't see how Synchronize is relevant?

On the other hand labeled and unlabeled look important https://developer.github.com/v3/activity/events/types/#pullrequestevent

Reply:
Hmm, I think you're right, `synchronized` is on updated code. Why would we need to watch for labels being added/removed?